signer/src/main.rs

```diff
@@ -68,48 +110,88 @@ async fn run_shutdown_signal_watcher(ctx: &impl Context) -> Result<(), Error> {
             let mut interrupt = tokio::signal::unix::signal(signal::unix::SignalKind::interrupt())?;
 
             tokio::select! {
+                // If the shutdown signal is received, we'll shut down the signal watcher
+                // by returning early; the rest of the components have already received
+                // the shutdown signal.
+                _ = term.wait_for_shutdown() => {
+                    tracing::info!("termination signal received, signal watcher is shutting down");
+                    return Ok(());
+                },
+                // SIGTERM (kill -15 "nice")
                 _ = terminate.recv() => {
                     tracing::info!(signal = "SIGTERM", "received termination signal");
                 },
+                // SIGHUP (kill -1)
                 _ = hangup.recv() => {
                     tracing::info!(signal = "SIGHUP", "received termination signal");
                 },
-                // Ctrl-C will be received as a SIGINT.
+                // Ctrl-C will be received as a SIGINT (kill -2)
                 _ = interrupt.recv() => {
                     tracing::info!(signal = "SIGINT", "received termination signal");
                 },
             }
         // Otherwise, we'll just listen for Ctrl-C, which is the most portable.
         } else {
-            tokio::signal::ctrl_c().await?;
-            tracing::info!(signal = "Ctrl+C", "received termination signal");
+            tokio::select! {
+                // If the shutdown signal is received, we'll shut down the signal watcher
+                // by returning early; the rest of the components have already received
+                // the shutdown signal.
+                Ok(_) = ctx.wait_for_shutdown() => {
+                    tracing::info!("termination signal received, signal watcher is shutting down");
+                    return Ok(());
+                },
+                _ = tokio::signal::ctrl_c() => {
+                    tracing::info!(signal = "Ctrl+C", "received termination signal");
+                }
+            }
+
         }
     }
 
     // Send the shutdown signal to the rest of the application.
     tracing::info!("sending shutdown signal to the application");
-    ctx.signal(SignerSignal::Shutdown)?;
+    term.signal_shutdown();
 
     Ok(())
 }
 
 /// Runs the libp2p swarm.
+#[tracing::instrument(skip(ctx))]
 async fn run_libp2p_swarm(ctx: &impl Context) -> Result<(), Error> {
-    // Subscribe to the signal channel so that we can catch shutdown events.
-    let mut signal = ctx.get_signal_receiver();
+    tracing::info!("initializing the p2p network");
 
-    // TODO(409): Add libp2p swarm initialization here.
+    // Convert the listen `Url`s from the config into `Multiaddr`s.
+    tracing::debug!("parsing listen addresses");
+    let mut listen_addrs: Vec<Multiaddr> = Vec::new();
+    for addr in ctx.config().signer.p2p.listen_on.iter() {
+        listen_addrs.push(addr.clone());
+    }
```

Comment:
I think this was just missed after the recent changes. But all of this can be simplified by just cloning the entire `listen_on` field, and the comment is off. Same thing for the `seed_addrs` variable below.